### PR TITLE
Fee map

### PR DESCRIPTION
**Pull request by roconnor-blockstream**
Add a jet for lookup up transaction fees by assetid.

### CODE REVIEW

File: C/primitive/elements/env.c [hunk 1]
```diff
@@ -515,7 +524,23 @@ extern transaction* elements_simplicity_mallocTransaction(const rawTransaction*
...
+     * Because the perm array entries are the same size as the feeOutputs array entries, it is safe to initialize one by one.
+     *
+     * In practical C implementations, the feeOutputs array entires are initalized to the same value as the perm array entries.
+     * In practical C implementations, this is a no-op, and generally compiliers are able to see this fact and eliminate this loop.
+     *
+     * We keep the loop in the code just to be pedantic.
+     */
+    for (size_t i = 0; i < numFees; ++i) {
+      static_assert(0 == offsetof(sigOutput, asset.data), "asset ID is not first field of sigOutput.");
+      feeOutputs[i] = (sigOutput*)(uintptr_t)(perm[i]);
```
Comment (real-or-random) [inline C/primitive/elements/env.c:561]: Why cast to `(uintptr_t)` first?

File: C/primitive/elements/env.c [hunk 2]
```diff
@@ -500,12 +501,20 @@ extern transaction* elements_simplicity_mallocTransaction(const rawTransaction*
     sha256_context ctx_outputSurjectionProofsHash = sha256_init(tx->outputSurjectionProofsHash.s);
     sha256_context ctx_outputsHash = sha256_init(tx->outputsHash.s);
     uint_fast32_t ix_fee = 0;
+
+    /* perm is a temporary array the same length (numFees) and size as feeOutputs.
+     * perm is used to initalize feeOutputs and is not used afterward.
+     * This makes it safe for perm to use the same memory allocation as feeOutputs.
+     */
+    static_assert(sizeof(const sha256_midstate*) == sizeof(sigOutput*), "Pointers (to sturctures) ought to have the same size.");
+    static_assert(alignof(const sha256_midstate*) == alignof(sigOutput*), "Pointers (to sturctures) ought to have the same alignment.");
```
Comment (real-or-random) [inline C/primitive/elements/env.c:530]: typo: sturctures (2x)

File: C/primitive/elements/env.c (hunk 1 above)
Comment (roconnor-blockstream) [reply C/primitive/elements/env.c:561]: This suppresses warning about both casting away `const` and the warning about the change in pointer alignment.

I'll insert a comment about this.

File: C/primitive/elements/env.c [hunk 3]
```diff
@@ -515,7 +524,23 @@ extern transaction* elements_simplicity_mallocTransaction(const rawTransaction*
...
+    simplicity_assert(ix_fee == numFees);
+    rsort(perm, numFees);
+
+    /* Initialize the feeOuputs array from the perm array.
+     * Because the perm array entries are the same size as the feeOutputs array entries, it is safe to initialize one by one.
+     *
+     * In practical C implementations, the feeOutputs array entires are initalized to the same value as the perm array entries.
+     * In practical C implementations, this is a no-op, and generally compiliers are able to see this fact and eliminate this loop.
+     *
+     * We keep the loop in the code just to be pedantic.
```
[the remark this answers is not in the document]
Comment (roconnor-blockstream) [reply C/primitive/elements/env.c:557]: Yes.

[the remark this answers is not in the document]
Comment (roconnor-blockstream) [reply C/primitive/elements/env.c:558]: I tried a couple of other avenues such as `rsort` taking a `void *` array or a `unsigned char *` array, but these changes didn't seem to make anything easier.  AFAICT, the C standard basically doesn't let you safely cast pointers to pointers into pointers to pointers of different types.  And indeed things get even less safer when if some of these types are non-structs.

**Comment by roconnor-blockstream** [comment]
@apoelstra I believe this PR is ready now.  You may want to run it through your CI first to see if it passes all your checks.

File: C/primitive/elements/primitive.h [hunk 4]
```diff
@@ -196,6 +197,7 @@ typedef struct sigInput {
 typedef struct transaction {
   const sigInput* input;
   const sigOutput* output;
+  const sigOutput* const * feeOutputs;
```
Comment (uncomputable) [inline C/primitive/elements/primitive.h:200]: The `transaction` type is opaque so FFI is not affected, right? We build `transaction` from `rawTransaction` and then pass the opaque struct to the Simplicity runtime, without caring about its contents.

Comment (roconnor-blockstream) [reply C/primitive/elements/primitive.h:200]: Yes. That is the intended API.

Comment (roconnor-blockstream) [reply C/primitive/elements/primitive.h:200]: The intention is that the transaction value can be reused between validating multiple Simplicity inputs within one transaction.

**Review by apoelstra — APPROVED** [review]
ACK 473f3d42fc14bc76c03915d0d44de4a5403ebacb Carefully read all C code, though despite my care most of the comments explaining clever stuff were about things I would have missed entirely if there were no comments

**Pull request merged**